firecracker-control/local.go

```diff
@@ -77,20 +80,35 @@ func newLocal(ic *plugin.InitContext) (*local, error) {
 	return &local{
 		vm:          make(map[string]*instance),
 		rootPath:    ic.Root,
+		publisher:   ic.Events,
 		findVsockFn: findNextAvailableVsockCID,
 	}, nil
 }
 
 // CreateVM creates new Firecracker VM instance
-func (s *local) CreateVM(ctx context.Context, req *proto.CreateVMRequest) (*proto.CreateVMResponse, error) {
-	id, err := s.makeID()
+func (s *local) CreateVM(ctx context.Context, req *proto.CreateVMRequest) (*empty.Empty, error) {
+	var (
+		id     = req.GetVMID()
+		logger = log.G(ctx).WithField("vm_id", id)
+	)
+
+	if id == "" {
+		return nil, errors.New("invalid VM ID")
+	}
+
+	// Make sure VM ID unique
+	if _, err := s.getVM(id); err == nil {
```

Comment:
I think we need this `getVM` call to be atomic with the `s.vm[id] = newInstance` call at the bottom of this function, otherwise it seems like there will still be race conditions for concurrent requests.

Reply:
Done